src/main/java/org/springframework/integration/aws/lock/DynamoDbLockRegistry.java

```diff
@@ -0,0 +1,552 @@
+
+	public void setLeaseDuration(long leaseDuration) {
+		this.leaseDuration = leaseDuration;
+	}
+
+	public void setHeartbeatPeriod(long heartbeatPeriod) {
+		this.heartbeatPeriod = heartbeatPeriod;
+	}
+
+	public void setRefreshPeriod(long refreshPeriod) {
+		this.refreshPeriod = refreshPeriod;
+	}
+
+	/**
+	 * Set the {@link Executor}, where is not provided then a default of
+	 * cached thread pool Executor will be used.
+	 * @param executor the executor service
+	 */
+	public void setExecutor(Executor executor) {
+		this.executor = executor;
+		this.executorExplicitlySet = true;
+	}
+
+	@Override
+	public void afterPropertiesSet() {
+		if (!this.dynamoDBLockClientExplicitlySet) {
+			AmazonDynamoDBLockClientOptions dynamoDBLockClientOptions =
+					AmazonDynamoDBLockClientOptions
+							.builder(this.dynamoDB, this.tableName)
+							.withPartitionKeyName(this.partitionKey)
+							.withSortKeyName(this.sortKeyName)
+							.withHeartbeatPeriod(this.heartbeatPeriod)
+							.withLeaseDuration(this.leaseDuration)
+							.build();
+
+			this.dynamoDBLockClient = new AmazonDynamoDBLockClient(dynamoDBLockClientOptions);
+		}
+
+		this.leaseDuration =
+				(long) new DirectFieldAccessor(this.dynamoDBLockClient)
+						.getPropertyValue("leaseDurationInMilliseconds");
+	}
+
+	@Override
+	public boolean isAutoStartup() {
+		return true;
+	}
+
+	@Override
+	public void start() {
+		if (!this.running.getAndSet(true)) {
+			if (!this.dynamoDBLockClientExplicitlySet) {
+				try {
+					this.dynamoDBLockClient.assertLockTableExists();
+					this.createTableLatch.countDown();
+					return;
+				}
+				catch (LockTableDoesNotExistException e) {
+					if (logger.isInfoEnabled()) {
+						logger.info("No table '" + this.tableName + "'. Creating one...");
+					}
+				}
+
+				CreateDynamoDBTableOptions createDynamoDBTableOptions =
+						CreateDynamoDBTableOptions
+								.builder(this.dynamoDB,
+										new ProvisionedThroughput(this.readCapacity, this.writeCapacity),
+										this.tableName)
+								.withPartitionKeyName(this.partitionKey)
+								.withSortKeyName(this.sortKeyName)
+								.build();
+
+				AmazonDynamoDBLockClient.createLockTableInDynamoDB(createDynamoDBTableOptions);
+			}
+		}
+
+		int i = 0;
+
+		boolean tableExists = false;
+
+		// We need up to one minute to wait until table is created on AWS.
+		while (i++ < 60) {
+			tableExists = this.dynamoDBLockClient.lockTableExists();
+			if (!tableExists) {
+				try {
+					// This is allowed minimum for constant AWS requests.
+					Thread.sleep(1000);
+				}
+				catch (InterruptedException e) {
+					ReflectionUtils.rethrowRuntimeException(e);
+				}
+			}
+			else {
+				break;
+			}
+		}
+
+		if (!tableExists) {
+			logger.error("Cannot describe DynamoDb table: " + this.tableName);
+		}
+
+		this.createTableLatch.countDown();
+	}
+
+	@Override
+	public void stop(Runnable callback) {
+		stop();
+		callback.run();
+	}
+
+	@Override
+	public void stop() {
+		this.running.set(false);
```

Comment:
Nothing to do on AWS?

Reply:
Well, agreed. Will move the `destroy()` logic here. When we stopped, we shouldn't schedule any heart beats to the AWS. Good catch! :+1: